### ISSUE TITLE
TypeError when deepcopying models that reference (and have sampled from) LogNormal

### ISSUE THREAD

**Issue by elvijs** [tensorflow/probability#944]
Our training procedure involves deepcopying a model. This fails when we have a parameter with `tfp.distributions.LogNormal` prior that is used early in training, but is subsequently disabled. The failure is `TypeError: can't pickle HashableWeakRef objects`. 

Note that `tfp.distributions.Normal` does not fail in the same way.

## To reproduce

I've attempted to trim away all irrelevant bits from our full training pipeline. It looks like the root cause relates to sampling from `tfp.distributions.LogNormal` (although anecdotally, we've seen similar issues with some bijectors).

**Minimal reproduction**
```python
import copy
import tensorflow as tf
import tensorflow_probability as tfp


def boom() -> None:
    m = tf.Module()
    m.prior = tfp.distributions.LogNormal(0.0, 1.0)  # error
    # m.prior = tfp.distributions.Normal(0.0, 1.0)  # no error

    @tf.function  # no error without the decorator
    def _sample() -> tf.Tensor:
        m.prior.sample()
        # it looks like the sampling pollutes the graph - if you comment out the line above,
        # then we don't explode

    _sample()
    # now that we've done some sampling at model initialisation, 
    # we actually want to disable the priors for training
    m.prior = None  # if you comment out this line, then we don't explode

    copy.deepcopy(m)


if __name__ == '__main__':
    boom()
```

**Stack trace**:
```Traceback (most recent call last):
  File "/home/elvijs/.PyCharmCE2019.2/config/scratches/scratch_7.py", line 27, in <module>
    boom()
  File "/home/elvijs/.PyCharmCE2019.2/config/scratches/scratch_7.py", line 23, in boom
    copy.deepcopy(m)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 215, in _deepcopy_list
    append(deepcopy(a, memo))
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 274, in _reconstruct
    y = func(*args)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 273, in <genexpr>
    args = (deepcopy(arg, memo) for arg in args)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 215, in _deepcopy_list
    append(deepcopy(a, memo))
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 274, in _reconstruct
    y = func(*args)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 273, in <genexpr>
    args = (deepcopy(arg, memo) for arg in args)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 280, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 150, in deepcopy
    y = copier(x, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 240, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 180, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 305, in _reconstruct
    key = deepcopy(key, memo)
  File "/home/elvijs/.virtualenvs/automl_res/lib/python3.7/copy.py", line 169, in deepcopy
    rv = reductor(4)
TypeError: can't pickle HashableWeakRef objects
```

## Expected behaviour

I expected no blow-up. 

## System information

* TensorFlow version: 2.2.0
* TensorFlow Probability version: 0.10.0
* Python version: 3.7.5
* Operating system: Ubuntu

**Comment by awav** [comment]
I have raised a similar issue, https://github.com/tensorflow/probability/issues/547

**Comment by csuter** [comment]
This is coming from the attempt to deep copy the `Exp` `Bijector` cache. The cache uses `weakref`s to avoid filling up memory over time with cached references to inputs/outputs that are no longer available, and these are not pickleable (nor, probably, should they be). We probably just need to tell copy/pickle to skip the cache; we'd welcome a PR if either of you have the bandwidth and know the right way to hook in to copy/pickle.

**Comment by elvijs** [comment]
Thanks for the quick response! Sounds like it should be a reasonably easy fix. I don't have the time right now, but will revisit this when I've got a bit more bandwidth

**Comment by emilyfertig** [comment]
This was resolved when we changed bijector caching from instance-level to global (so it's fixed in TFP 0.12 and later).

**Comment by elvijs** [comment]
> This was resolved when we changed bijector caching from instance-level to global (so it's fixed in TFP 0.12 and later).

Awesome, thanks!
